adafruit_pioasm.py

```diff
@@ -58,18 +68,65 @@ def __init__(self, text_program: str, *, build_debuginfo: bool = False) -> None:
         wrap = None
         wrap_target = None
         offset = -1
+        pio_version = 0
+        fifo_type = "auto"
+        mov_status_type = None
+        mov_status_n = None
+        in_count = None
+        in_shift_right = None
+        auto_push = None
+        push_threshold = None
+        out_count = None
+        out_shift_right = None
+        auto_pull = None
+        pull_threshold = None
+        set_count = None
+
+        def require_before_instruction():
+            if len(instructions) != 0:
+                raise RuntimeError(f"{words[0]} must be before first instruction")
+
+        def require_version(required_version, instruction):
+            if pio_version < required_version:
+                raise RuntimeError(
+                    f"{instruction} requires .pio_version {required_version}"
+                )
+
+        def int_in_range(arg, low, high, what, radix=0):
+            result = int(arg, radix)
+            if low <= result < high:
+                return result
+            raise RuntimeError(
+                f"{what} must be at least {low} and no greater than {high}, got {result}"
+            )
+
+        def parse_rxfifo_brackets(arg, fifo_dir):
+            require_version(1, line)
+            if (  # pylint: disable=consider-using-in
+                fifo_type != "putget" and fifo_type != fifo_dir
+            ):
+                raise RuntimeError(
+                    f"FIFO must be configured for '{fifo_dir}' or 'putget' for {line}"
+                )
+            if arg.endswith("[y]"):
+                return 0b1000
+            return int_in_range(arg[7:-1], 0, 8, "rxfifo index")
```

Comment:
I think this is inverted. bit 3 is high when the index is in the instruction. So y is 0b0000.

The included index is only 2 bit too so the range is 0, 4.

Reply:
Yes, you were right.

I added to the tests ~1200 instructions, "substantially all" of them, generated by script invoking sdk pioasm. Now all 1200 of them are parsed by adafruit_pioasm and give the same instruction word (this does not test all side set & delay values though)

Given this bug it's a bit of a surprise that my rxfifo-using program worked, I wonder why/how. It was using an invalid `mov ,rxfifo` instruction! interesting.